src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerActor.scala

```diff
@@ -125,13 +131,18 @@ private class TaskTrackerActor(
         val op = TaskOpProcessor.Operation(deadline, sender(), taskId, taskStateOp)
         updaterRef.forward(TaskUpdateActor.ProcessTaskOp(op))
 
-      case msg @ TaskTrackerActor.TaskUpdated(task, ack) =>
-        becomeWithUpdatedApp(task.appId)(task.taskId, newTask = Some(task))
-        ack.sendAck()
-
-      case msg @ TaskTrackerActor.TaskRemoved(taskId, ack) =>
-        becomeWithUpdatedApp(taskId.appId)(taskId, newTask = None)
-        ack.sendAck()
+      case msg @ TaskTrackerActor.StateChanged(change) =>
+        val taskAfterChange = change.stateChange match {
+          case TaskStateChange.Update(task, _) => Some(task)
+          case _                               => None
+        }
+        becomeWithUpdatedApp(change.appId)(change.taskId, newTask = taskAfterChange)
+        val originalSender = sender()
+
+        import context.dispatcher
+        updateStepProcessor.process(change).map { _ =>
+          originalSender ! (())
```

Comment:
you need to handle the failure case here. (and I'd prefer foreach for side effects)

I am not sure how to deal with errors here. I think for now, we can log errors and proceed anyway, since we use "ContinueOnError" on the steps anyway and are apparently fine with that strategy.